portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,34 +24,51 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import javax.ws.rs.core.MediaType;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   private Gson gson = new Gson();
   private DatastoreInterface datastoreInterface = new DatastoreInterface();
+  private UserService userService = UserServiceFactory.getUserService();
+
+  private class DataPackage {
+      private boolean isUserLoggedIn;
+      private List<Comment> comments;
+      DataPackage(){
+          this.isUserLoggedIn = DataServlet.this.userService.isUserLoggedIn();
+          this.comments = DataServlet.this.datastoreInterface.getComments();
+      }
+
+  }
```

Comment:
This is really closely coupled with the surrounding class.
This means it has to know both the interface of it instead of being as simple and dumb as possible.
Also executing logic like this in the constructor can be dangerous because it might introduce transitive dependencies that you aren't aware of. (in larger projects where you might not know the exact impact of calling such a method)

Instead, consider making this a dumb Pojo which receives the loggedin flag and comments list via the constructor.

Some resources aroud this are: 
https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)#Low_coupling
https://en.wikipedia.org/wiki/SOLID

Reply:
then the actual fetching of comments can happen in your `dataToJson` method